Python/bytecodes.c

```diff
@@ -3945,16 +3957,22 @@ dummy_func(
 
         op(_SET_IP, (--)) {
             TIER_TWO_ONLY
-            frame->prev_instr = ip_offset + oparg;
+            frame->instr_ptr = ip_offset + oparg;
         }
 
         op(_SAVE_CURRENT_IP, (--)) {
-            TIER_ONE_ONLY
-            frame->prev_instr = next_instr - 1;
+            #if TIER_ONE
+            assert(frame->next_instr_offset == 0);
+            frame->next_instr_offset = (uint16_t)(next_instr - frame->instr_ptr);
+            #endif
+            #if TIER_TWO
+            frame->next_instr_offset = oparg;
+            #endif
         }
 
         op(_EXIT_TRACE, (--)) {
-            frame->prev_instr--;  // Back up to just before destination
+            TIER_TWO_ONLY
+            frame->next_instr_offset = 0;  // Dispatch to frame->instr_ptr
```

Comment:
Again, this should be needed. In fact it might be wrong, if `next_instr_offset` has been set by an earlier uop.
E.g., `... POP_FRAME; EXIT_TRACE`

Reply:
Again, without this line I got test failures (in `test_capi.test_misc.TestUops`, without `-Xuops` even). But you're also right that on the face of it, it looks wrong. Like above, I'm not sure how to reason about all this. It's possible that that example doesn't occur because `_EXIT_TRACE` is always preceded by `SET_IP`.